### PR TITLE
ADR-04: reactor stub command

### LINKED ISSUE #15: ADR-04: generating reactor stub
... following on the Issue #14

### PR DESCRIPTION
**Pull request by ivan-gavran**
[rendered document](https://github.com/informalsystems/atomkraft/blob/ivan/adr04Reactor/docs/src/adr/04adr-reactor-stub.md)

When finished, this PR will close #15

### CODE REVIEW

File: docs/src/adr04-reactor-stub/adr04-reactor-stub.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+The reason for updating the config file is to give users a way to silently refer to the most recent model and reactor files.
+
+
+### Stub content
+The code assumes the access to the `config` dictionary, which has access to everything written in the `.atomkraft` file.
+Furthermore, it assumes imported `Testnet` from `cosmos_net.pytest`.
+
+ The stub should include:
+
+ - a stub for the testnet initialization function, based on the value of the `binary` that was put into `.atomkraft` by the command `atomkraft init`.
```
Comment (ivan-gavran) [inline docs/src/adr04-reactor-stub/adr04-reactor-stub.md:45]: @rnbguy , would like agree with you about the artifacts of the init function. E.g., will it save only the binary name, or all the values necessary to span the default testnet.

If former, then we should figure out where to keep default values for familiar binaries.

File: docs/src/adr/04adr-reactor-stub.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+## Implementation
+
+ The `ReactorRoom` class implements the desired behavior.
+ Its member functions are:
+  - `generate_reactor(actions_list, variables_list, stub_file_path=None)`:
+  generates the stub, containing one `@step` function for each action from the `action_list`,
+  and each of these functions takes variables from the `variables_list` as arguments.
+  The reactor stub is created at `stub_file_path` name location
+  (if it is `None`, a default location is used).
+  - `check_reactor(trace, keypath, reactor=None)`: it checks if the `reactor` (default reactor if `None`) is compatible with the `trace`. A reactor is compatible with a trace if each action appearing in `trace` is matched with a function in `reactor`. The `keypath` argument defines which field of the ITF trace contains the action.
```
Comment (andrey-kuprianov) [inline docs/src/adr/04adr-reactor-stub.md]: the problem with having the function `check_reactor()` is that as its user (the `Execute` component), if I don't have a reactor (the end user didn't provide me such argument), I don't have anything to check. I can only kindly _ask_ to give me a reactor for the given trace; thus my suggestion to use `get_trace_reactor(trace, reactor = None)`. In the above version, and additional function `get_reactor()` would be needed anyway, in case there is only a default reactor.

Comment (ivan-gavran) [reply docs/src/adr/04adr-reactor-stub.md]: I had a similar comment on your PR, but really suggesting the opposite.
My reasoning was that I will give a default reactor if it is not provided (that's why it has a default argument `None`).
If I do the same with `get_trace_reactor`, then providing the `reactor` doesn't make sense (because I would then always give the same path back). This is all conditioned on the assumption that we are dealing with paths here (and not some other kinds of objects).

I modified this passage now to be more precise about the paths assumption and about what is the return value (that is completely missing now)

Comment (andrey-kuprianov) [reply docs/src/adr/04adr-reactor-stub.md]: Ok, I see:) so our functions, according to your description do basically the same thing. The problem with the naming of your function is that it violates one of the basic assumptions: `check_` functions always return a boolean. I suggest we rename it to `get_`; let's discuss online if you disagree

File: docs/src/adr/04adr-reactor-stub.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+### Stub content
+
+ The stub should include:
+
+
+  - a stub for the `chain_testnet` initialization function
+  ```python
+  @pytest.fixture(scope="session")
+  def chain_testnet(num_validators=3, num_accounts=3, account_balance=1000, validator_balance=100):
```
Comment (ivan-gavran) [inline docs/src/adr/04adr-reactor-stub.md]: What would be the best default values here?

File: docs/src/adr/04adr-reactor-stub.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+
+### Stub content
+
+ The stub should include:
+
+
+  - a stub for the `chain_testnet` initialization function
+  ```python
+  @pytest.fixture(scope="session")
```
Comment (ivan-gavran) [inline docs/src/adr/04adr-reactor-stub.md]: @rnbguy , is it fair to assume that I'll get all these values  (`node_config`, `genesis_config`) from the config file?

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v2.3.0
+    hooks:
+    -   id: check-yaml
+    -   id: end-of-file-fixer
+    -   id: trailing-whitespace
+-   repo: https://github.com/psf/black
+    rev: 21.12b0
+    hooks:
+    -   id: black
```
Comment (andrey-kuprianov) [inline .pre-commit-config.yaml]: Why is this file included into the PR? Doesn't really seem to rightfully belong here...

File: docs/src/adr/04adr-reactor-stub.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            "api.enabled-unsafe-cors": True,
+            "minimum-gas-prices": f"0.10{denom}",
+            "rosetta.enable": False,
+        },
+        "config/config.toml": {
+            "instrumentation.prometheus": False,
+            "p2p.addr_book_strict": False,
+            "p2p.allow_duplicate_ip": True,
+        },
+    }
```
Comment (andrey-kuprianov) [inline docs/src/adr/04adr-reactor-stub.md]: To me it seems that all those should be read from the Atomkraft config files

**Review by andrey-kuprianov — CHANGES_REQUESTED** [review]
Looks good, but some changes are needed, I believe... also a review from @rnbguy is needed.
